src/psyclone/psyir/transformations/intrinsics/sum2code_trans.py

```diff
@@ -0,0 +1,361 @@
+            raise TransformationError(
+                f"Only real and integer types supported for array "
+                f"'{array_ref.name}', but found '{array_intrinsic.name}'.")
+
+    def apply(self, node, options=None):
+        '''Apply the SUM intrinsic conversion transformation to the specified
+        node. This node must be a SUM Operation which is converted to
+        equivalent inline code.
+
+        :param node: a Sum Operation. This could be a UnaryOperation, \
+            BinaryOperation or NaryOperation depending on how many \
+            arguments it has.
+        :type node: :py:test:`psyclone.psyir.nodes.operation.UnaryOperation` \
+            | :py:test:`psyclone.psyir.nodes.operation.BinaryOperation` | \
+            :py:test:`psyclone.psyir.nodes.operation.NaryOperation`
+        :param options: options for the transformation.
+        :type options: Optional[Dict[str,str]]
+
+        '''
+        self.validate(node)
+
+        array_ref, dimension_ref, mask_ref = get_args(node)
+
+        # Determine the literal value of the dimension argument
+        dimension_literal = None
+        if not dimension_ref:
+            # there is no dimension argument
+            pass
+        elif isinstance(dimension_ref, Literal):
+            dimension_literal = dimension_ref
+        elif (isinstance(dimension_ref, Reference) and
+              dimension_ref.symbol.is_constant):
+            dimension_literal = dimension_ref.symbol.constant_value
+        # else exception is handled by the validate method.
+
+        # Determine the dimension and extent of the array
+        ndims = None
+        if len(array_ref.children) == 0:
+            if not array_ref.symbol.is_array:
+                # Exception handled by validate method
+                pass
+            ndims = len(array_ref.symbol.shape)
+
+            loop_bounds = []
+            for idx, shape in enumerate(array_ref.symbol.shape):
+                if shape in [ArrayType.Extent.DEFERRED,
+                             ArrayType.Extent.ATTRIBUTE]:
+                    # runtime extent using LBOUND and UBOUND required
+                    lbound = BinaryOperation.create(
+                        BinaryOperation.Operator.LBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    loop_bounds.append((lbound, ubound))
+                elif isinstance(shape, ArrayType.ArrayBounds):
+                    # array extent is defined in the array declaration
+                    loop_bounds.append(shape)
+                #  else: Exception is handled by validate method
+        else:
+            # This is an array reference
```

Comment:
Not necessarily unless is checked in the validate

Reply:
Line 186 in the validate method guarantees this.